projects/solar/test/test_relay_fsm.c

```diff
@@ -0,0 +1,147 @@
+#include "log.h"
+#include "ms_test_helper_can.h"
+#include "ms_test_helpers.h"
+#include "solar_events.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#define IRRELEVANT_EVENT NUM_SOLAR_RELAY_EVENTS
+
+#define PROCESS_RELAY_FSM_EVENT(storage, e)    \
+  ({                                           \
+    MS_TEST_HELPER_AWAIT_EVENT(e);             \
+    relay_fsm_process_event(&(storage), &(e)); \
+    MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();   \
+  })
+
+static const GpioAddress s_test_relay_pin = { .port = GPIO_PORT_A, .pin = 6 };
+
+static CanStorage s_can_storage;
+static RelayFsmStorage s_storage;
+
+void setup_test(void) {
+  event_queue_init();
+  drv120_relay_init(&s_test_relay_pin);
+  TEST_ASSERT_OK(relay_fsm_init(&s_storage));
+}
+void teardown_test(void) {}
+
+// Test that the relay starts closed and is changed in response to command functions.
+void test_relay_transitions(void) {
+  Event e = { 0 };
+  bool closed;
+
+  // starts closed
+  drv120_relay_get_is_closed(&closed);
+  TEST_ASSERT_TRUE(closed);
+
+  // closed -> open
+  TEST_ASSERT_OK(relay_fsm_open());
+  PROCESS_RELAY_FSM_EVENT(s_storage, e);
+  drv120_relay_get_is_closed(&closed);
+  TEST_ASSERT_FALSE(closed);
+
+  // open -> closed
+  TEST_ASSERT_OK(relay_fsm_close());
+  PROCESS_RELAY_FSM_EVENT(s_storage, e);
+  drv120_relay_get_is_closed(&closed);
+  TEST_ASSERT_TRUE(closed);
+
+  // closed -> open
+  TEST_ASSERT_OK(relay_fsm_open());
+  PROCESS_RELAY_FSM_EVENT(s_storage, e);
+  drv120_relay_get_is_closed(&closed);
+  TEST_ASSERT_FALSE(closed);
+
+  // open still (should be no-op)
+  TEST_ASSERT_OK(relay_fsm_open());
+  PROCESS_RELAY_FSM_EVENT(s_storage, e);
+  drv120_relay_get_is_closed(&closed);
+  TEST_ASSERT_FALSE(closed);
+
+  // open -> closed
+  TEST_ASSERT_OK(relay_fsm_close());
+  PROCESS_RELAY_FSM_EVENT(s_storage, e);
+  drv120_relay_get_is_closed(&closed);
+  TEST_ASSERT_TRUE(closed);
+
+  // closed still (should be no-op)
+  TEST_ASSERT_OK(relay_fsm_close());
+  PROCESS_RELAY_FSM_EVENT(s_storage, e);
+  drv120_relay_get_is_closed(&closed);
+  TEST_ASSERT_TRUE(closed);
+}
+
+// Test that the FSM doesn't respond to irrelevant events.
+void test_irrelevant_events(void) {
+  Event e = { 0 };
+  bool closed;
+
+  // reset to open
+  TEST_ASSERT_OK(relay_fsm_open());
+  PROCESS_RELAY_FSM_EVENT(s_storage, e);
+  drv120_relay_get_is_closed(&closed);
+  TEST_ASSERT_FALSE(closed);
+
+  e.id = IRRELEVANT_EVENT;
+  relay_fsm_process_event(&s_storage, &e);
+  drv120_relay_get_is_closed(&closed);
+  TEST_ASSERT_FALSE(closed);  // not changed
+}
+
+// Test that |relay_fsm_process_event| returns whether there was a state change.
+void test_relay_fsm_process_event_return_value(void) {
+  // force to closed at start - valid in case we don't initialize to closed in the future
+  Event e = { 0 };
+  TEST_ASSERT_OK(relay_fsm_close());
+  PROCESS_RELAY_FSM_EVENT(s_storage, e);
+
+  e.id = IRRELEVANT_EVENT;
+  TEST_ASSERT_FALSE(relay_fsm_process_event(&s_storage, &e));
+
```

Comment:
should probably verify that it is indeed closed here using `drv120_relay_get_is_closed`

Reply:
I feel like this is fine since this test is just supposed to test `relay_fsm_process_event`'s return value, the other tests already test the relay state (including with respect to irrelevant events).

Reply:
Fair enough, you only need this test because you don't check the return value in the other tests ... maybe extending the macro so you assert it might be worth?

Reply:
Hm, extending the macro would probably make the other tests less clear since we'd have a seemingly unrelated bool parameter. We'd also need two macros, one checking and one not, since we don't want to check the return value when resetting the relay to open, since that would introduce a dependency on the initial state of the relay which I only want in the `test_starts_closed` test. It's probably fine having a test just to check the return value.

Reply:
The assumption is already there, but it's up to you